### ISSUE TITLE
RouterService.urlFor causes error in non-application tests.

### ISSUE THREAD

**Issue by apellerano-pw** [emberjs/ember.js#16831]
### Steps to reproduce
1. Use `urlFor()` in a component and output the value in the template
1. Attempt to render the component in an integration/rendering test

```
TypeError: Cannot read property 'generate' of undefined
    at Class.generate (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:33524:59)
    at Class.urlFor (https://cdn.jsdelivr.net/npm/ember-source@3.2.2/dist/ember.debug.js:31371:48)
```

### Expected behavior
The test doesn't crash.

Since `autoboot=false` is in use it makes sense that there's no Router for RouterService to work with. However, I would expect it to return empty data rather than crash. Similar to how `{{link-to}}` behaves in the same scenario.

### Twiddle
If you click Run Tests you'll see the exception.
https://ember-twiddle.com/bc59e11d525fc3548519f66250173a32?openFiles=my-component.component.js%2C

**Comment by pixelhandler** [comment]
> Since autoboot=false is in use it makes sense that there's no Router for RouterService to work with. However, I would expect it to return empty data rather than crash. Similar to how {{link-to}} behaves in the same scenario.

What is the value of using `autoboot=false` in your use case? I'm not sure what that does.

Also what about stubbing the router service in your test?

**Comment by pixelhandler** [comment]
@apellerano-pw Thanks for the explanation :)

**Comment by pixelhandler** [comment]
@apellerano-pw did you try stubbing the router in your test?

**Comment by apellerano-pw** [comment]
It could be a workaround, but it would be stubbing ember functionality and not my own.

I think part of the problem is that the RouterService is not just used for transitions, but also as a url generator. It makes no sense for the transition stuff to be available in a unit or rendering test, but it does make sense for url generation to exist

**Comment by rwjblue** [comment]
Yes, I totally agree that this is a bug. During `setupRenderingTests` type tests, I would absolutely want `this.router.urlFor(...)` and other methods to function properly.

The issue isn't really with `autoboot=false` though, inside `setupTest` / `setupRenderingTest` type tests you do have a fully booted application, **but** we specifically avoid starting routing as part of that setup. We probably should make the router service a tad more resilient in these cases. As a temporary work around @apellerano-pw you could do `this.owner.setupRouter()` in your test setup.

I'd gladly accept a PR to call `owner.setupRouter()` from any method that inherently requires the underlying `router:main` in the router service.

**Comment by dhuang612** [comment]
I would like to work on this

**Comment by sly7-7** [comment]
@rwjblue Looking at the router service code, I think this issue could be closed, as the router service calls `router:main.setupRouter()` the first time its `_router` getter is hit.

**Comment by rwjblue** [comment]
Yep, you are totally right. This should be fixed in 3.24 and 3.25 latest patch releases.
